TrailBlaze/Networking/PaceImprovementTracker.m

```diff
@@ -63,25 +66,47 @@ - (void) paceTracker: (NSArray *) polylinePoints userLocation: (CLLocation *) us
         
     }];
-    __block int i = 1;
-    __block NSArray *nextPoints = [polylinePoints subarrayWithRange:NSMakeRange(i, 2)];
-    __block NSDate *startDate = [NSDate date];
+     self.nextPoints = [polylinePoints subarrayWithRange:NSMakeRange(self.i, 2)];
     
-    while (![nextPoints[1] isEqual:polylinePoints[-1]]) {
-        if ([self passedPoint:nextPoints currentLocation:userLocation]) {
-            NSNumber *previousPace = bestPaces[nextPoints[0]] ;
+    if (![self.nextPoints[1] isEqual:polylinePoints[-1]]) {
+        if ([self passedPoint:self.nextPoints currentLocation:userLocation]) {
+            NSNumber *previousPace = bestPaces[self.nextPoints[0]] ;
             __block NSNumber *currentPace;
             NSDate *endDate = [NSDate date];
-            [self.pedometer queryPedometerDataFromDate:startDate toDate:endDate withHandler:^(CMPedometerData * _Nullable pedometerData, NSError * _Nullable error) {
+            [self.pedometer queryPedometerDataFromDate:self.startDate toDate:endDate withHandler:^(CMPedometerData * _Nullable pedometerData, NSError * _Nullable error) {
                 currentPace = pedometerData.averageActivePace;
+                self.currentPacesDictionary[polylinePoints[self.i-1]] = currentPace;
                 
                 // make delegate to handle
-                [self paceCompare:previousPace currentIntervalPace:currentPace pointsForInterval:@[polylinePoints[i-1], polylinePoints[i]] ];
-                i += 1;
-                nextPoints = [polylinePoints subarrayWithRange:NSMakeRange(i, 2)];
-                startDate = endDate;
+                [self paceCompare:previousPace currentIntervalPace:currentPace pointsForInterval:@[polylinePoints[self.i-1], polylinePoints[self.i]] ];
+                self.i += 1;
+                self.nextPoints = [polylinePoints subarrayWithRange:NSMakeRange(self.i, 2)];
+                self.startDate = endDate;
             }];
         } 
+    } else {
+        [self saveImprovedPaceDictionary:self.runObject];
     }
 }
+
+- (void) saveImprovedPaceDictionary: (PFObject *) runObject{
+    NSArray *currentPacesArray = [self.currentPacesDictionary allValues];
+    NSArray *bestPacesArray = [self.bestPacesDictionary allValues];
+    double currentPaceTotalToAverage = 0;
+    double bestPaceTotalToAverage = 0;
+    
+    for (int i = 0; i<bestPacesArray.count; i++) {
+        currentPaceTotalToAverage += [currentPacesArray[i] doubleValue];
+        bestPaceTotalToAverage += [bestPacesArray[i] doubleValue];
+    }
+    currentPaceTotalToAverage/=bestPacesArray.count;
+    bestPaceTotalToAverage/=bestPacesArray.count;
+    
+    if (currentPaceTotalToAverage >= bestPaceTotalToAverage) {
+        [runObject setValue:_currentPacesDictionary forKey:@"pacesDict"];
+        [runObject setValue:[NSString stringWithFormat:@"%f", currentPaceTotalToAverage] forKey:@"overallAveragePace"];
+        [runObject save];
```

Comment:
i'm feeling it might be too much for this component to be owning and updating this run object. a better way to handle this might be via a Delegate that has a callback like `didUpdateBestPace:overallAverPace:`, where the delegate owns the runObject. 

just another separation of concerns thing where this component should only deal with pace related things and not the overall run.